RFC.md

```diff
@@ -105,345 +63,485 @@ Thus:
 
 #### Initial device
-When a device is initially added to an account the device has to generate all relevant key material for the user account and provision those on the server. 
-
-First, the client has to generate the relevant key material:
+When a device is initially added to an account the device has to generate all relevant key material for the user account and provision those on the server:
 
-1. Client has to generate a new X.509 certificate request and private key.
-   1. CN of X.509 certificate must be set to the currently logged-in User ID
+1. Client has to generate a new X.509 certificate request and private key (TODO: check if we want/need to use PEM, otherwise base64 encode)
+  1. CN of X.509 certificate must be set to the currently logged-in User ID
 2. Client uploads the X.509 certificate request to the server by sending the certificate request URL encoded as parameter `csr` to `/ocs/v2.php/apps/end_to_end_encryption/api/v1/public-key`.
-3. Server issues a certificate if the CN matches the current user ID.
-4. Server returns the issued certificate.
-5. Client stores the private and the certificate in the keychain of the device.
+3. Server issues a certificate if the CN matches the current user ID
+4. Server returns the issued certificate
+5. Client stores the private key and the certificate in the keychain of the device
 
 In a second step, the private key will be stored encrypted on the server to simplify the addition of further devices:
 
-1. Client generates a 12 word long mnemonic from the english BIP-0039 word list. The word list contains 2048 words, thus resulting in 2048^12 possible key combinations.
-2. Client encrypts the private key using AES/GCM/NoPadding as cipher (256 bit key size) and uses PBKDF2WithHmacSHA1 as key derivation, as password the in step 1 generated mnemonic is used. The needed salt and initizialization vector is appended to the cipher text with base 64 encoded "|":  encryptedAndEncryptedBytes + "fA==" + encodedIV + "fA==" + encodedSalt
-3. Client uploads the encrypted X.509 private key to the server by sending the encrypted private key URL encoded as parameter `privateKey` to `/ocs/v2.php/apps/end_to_end_encryption/api/v1/private-key`. 
-4. The mnemonic is displayed to the user and the user is asked to store a copy in a secure place. For convenient reasons the mnemonic can be displayed with whitespaces, but the string for encrypting/decrypting must have no whitespaces and be lowercase.
-5. The mnemonic is stored in the keychain of the device (ideally with spaces so it can be shown more readable).
+1. Client generates a 12 word long mnemonic from the English BIP-0039 word list. The word list contains 2048 words, thus resulting in 2048^12 possible key combinations.
+2. Client encrypts the private key using AES/GCM/NoPadding as cipher (256 bit key size) and uses PBKDF2WithHmacSHA1 as key derivation with 1024 iterations, as password the in step 1 generated mnemonic is used. 
```

Comment:
Is there a specific reason for using SHA1 here? While this PBKDF2 version is generally fine the mnemonic will most likely be hashed with SHA1 first in this case as it is probably too long (average word size in bip-0039 is 5.4 such that the mnemonic can be larger than 64 bytes), see [1].

[1] https://tools.ietf.org/html/rfc2898#appendix-B.1